tdl2db.py

```diff
@@ -1,78 +1,181 @@
-                              sep = '\t',
-                        file=log)
-            except Exception as e:
-                print("Unable to parse tdl for {}, see log for details".format(os.path.join(root, f)),
-                      file=sys.stderr)
-                print("Unable to parse tdl for {}".format(os.path.join(root, f)),
-                       file=log)
-                if hasattr(e, 'message'):
-                    print(e.message, file=log)
+#grammar = '/home/bond/svn/erg-mo/educ.tdl'
+tdls = []
+hierarchy = []  #(child, parent)
+types = dd(list)
+les = {}
+
+
+def readgrm (grammarfile, tdls, types, hierarchy, les):
+    print("FILE", grammarfile)
+    path = Path(grammarfile)
+    base = path.parent
+    for event, obj, lineno in tdl.iterparse(grammarfile):
+        if  event == "LineComment":
+            continue
+        #print (event, obj, lineno)
+        # if event == "BeginEnvironment":
+        #     status = getattr(obj, 'status', 'type')
+        #     print ("Environment is", status)
+        if event == "EndEnvironment":
+            status = getattr(obj, 'status', 'type')
+            for entry in obj.entries:
+                #print('ENTRY', entry, status)
+                if isinstance(entry,tdl.FileInclude):
+                    path = Path(base, entry.path).with_suffix('.tdl')
+                    if path.is_file():
+                        process_type(str(base), str(path), status, tdls, hierarchy, les)
+                    else:
+                        print('INCLUDED FILE NOT FOUND: {!s}'.format(path))
+                else:
+                    print('what to do with', entry.status, file=log)
+
+def  process_type(base, path, status, tdls, hierachy, les):
+    if 'root' in path:
+        status = 'root'
+    elif 'parse-nodes' in path:
+        status = 'labels'
+
+    print(f"Processing types in {path} as {status}")
+    for event, obj, lineno in tdl.iterparse(path): # assume utf-8
+        #print(lineno, event, sep = '\t')
+        if event in ['TypeDefinition',  'TypeAddendum',
+                     'LexicalRuleDefinition']:
+            # if obj.documentation(): ### The tdl has a docstring
+            #     descript,exes,nams= ltdb.munge_desc(obj.identifier,obj.documentation())
+            #     obj.docstring=None
+            # else:
+            # get the parents
+            parents = [c for c in obj.conjunction.types()]
+            if status == 'lex-entry':
+                if len(parents) != 1:
+                    print ("LE has non unique parent", obj.identifier, parents)
                 else:
-                    print(str(e), file=log)
+                    ### (lex-type, docstring)
+                    ### fixme get orth, pred, altpred
+                    orths = obj.conjunction.get('ORTH', default=None)
+                    try:
+                        orth=' '.join([str(s) for s in orths.values()])
+                    except:
```

Comment:
Blanket exception. So you expect this to fail if `orths` is `None`? How about something like this (not tested):

```python
if 'ORTH' in obj:
    orths = ' '.join(map(str, obj['ORTH'].values()))
else:
    orth = ''
    print('No Orthography', obj.indentifier, sep='\t', file=log)
...
```

If this generates an error, perhaps it's a bug in PyDelphin or some other case we need to handle separately.